src/player/mod.rs

```diff
@@ -0,0 +1,159 @@
+use std::sync::Arc;
+
+use librespot::{
+    core::{
+        config::SessionConfig,
+        session::Session,
+        cache::Cache, SpotifyId,
+    },
+    playback::{
+        audio_backend,
+        config::{AudioFormat, PlayerConfig},
+        mixer::NoOpVolume,
+        player::{Player, PlayerEvent}
+    }
+};
+
+use crate::{storage::local::LocalStorage, track::Track};
+
+pub mod action;
+pub mod interface;
+mod state;
+
+async fn get_session_with_cached_credentials(cache_location: std::path::PathBuf) -> anyhow::Result<Session> {
+    let session_config = SessionConfig::default(); 
+    let cache = Cache::new(Some(cache_location), None, None, None)?;
+    
+    let session = Session::new(session_config, Some(cache.clone())); 
+    if let Some(cached_cred) = cache.credentials() {
+        match session.connect(cached_cred,true).await {
+            Ok(_) => Ok(session), 
+            Err(err) => {
+                tracing::error!("{}", err); 
+                Err(anyhow::anyhow!("invalid session credentials"))
+            }
+        }
+    } else {
+        Err(anyhow::anyhow!("invalid credentials in cache, please regenerate them with get_token"))
+    }
+}
+
+pub async fn get_player(cache_location: std::path::PathBuf) -> anyhow::Result<Arc<Player>> {
+    let session = get_session_with_cached_credentials(cache_location).await?;
+
+    let player_config = PlayerConfig::default();
+    let audio_format = AudioFormat::default();
+    let backend = audio_backend::find(None).unwrap();
+    let player = Player::new(player_config, session, Box::new(NoOpVolume), move || {
+        backend(None, audio_format)
+    });
+    Ok(player)
+}
+
+
+pub struct SpotifyPlayer<LocalStorage> {
+    player: Arc<Player>,
+    interface: Arc<interface::PlayerInterface<LocalStorage>>,
+    rx: flume::Receiver<action::PlayerAction>
+}
+
+
+
+impl SpotifyPlayer<LocalStorage> {
+    pub async fn new(
+        player: Arc<Player>,
+        interface: Arc<interface::PlayerInterface<LocalStorage>>,
+        rx: flume::Receiver<action::PlayerAction>
+    ) -> Self {
+        Self {
+            player,
+            interface,
+            rx
+        }
+    }
+
+    async fn play_track(&self, track: Track) {
+        let track_id = track.id;
+        if let Some(track_id) = track_id {
+            let spotify_track_id = SpotifyId::from_uri(&track_id.to_string()).unwrap();
```

Comment:
@shurikv Theoretically this should never fail because this value is being fetched from Spotify API. I have added validation step for the user provided input. However, this you are right and it is better to return Result here.